nm: Retry when NetworkManager starting up

cathay4t opened the pull request:
nm: Retry when NetworkManager starting up

When OS boot up with `nmstate.service` enabled, nmstate will fail as
NetworkManager daemon is not fully setup the DBUS interface yet.

The `nmstate.service` has `After=NetworkManager.service`. Considering
`NetworkManager.service` is dbus type systemd service, it is guaranteed
that when `nmstate.service` starts, the DBUS interface is acquired by
NetworkManager. The failure is
`org.freedesktop.DBus.Error.UnknownMethod` as NetworkManager has not
finished the DBUS interface setup yet.

The `org.freedesktop.DBus.Error.UnknownMethod` is too generic, we cannot
assume it as daemon not started with checking the error string.
Considering the DBUS error string is not an API, we will not parse that
error message. Instead, we change top level retry to cover on querying
failures in `apply()`. The retry has been expend to twice with 2 seconds
interval(was 1 second interval).

Do not want to create complex integration test case for this corner case.
Manually tested.

bgilbert commented:
I don't think `-wait-online` is the right fix here.  That service waits for NM to think it has a valid network connection, but all we actually need is for the NM D-Bus object to be available.  And in some environments NM might never decide that it's online without the configuration nmstate is going to provide.

cathay4t commented:
> 

Thanks! Let me investigate more.

cathay4t commented:
New patch/approach uploaded. Please review.

cathay4t commented:
V3 patch uploaded with `ping()` check removed as `After=NetworkManager.service` has guarantee on DBUS interface been acquired.

bgilbert commented on rust/src/lib/query_apply/net_state.rs:
```diff
@@ -172,33 +175,27 @@ impl NetworkState {
         let timeout = self.timeout.unwrap_or(DEFAULT_ROLLBACK_TIMEOUT);
         // NM might have unknown race problem found by verify stage,
         // we try to apply the state again if so.
```
Looks like a stale comment?

cathay4t replied:
Updated. Thanks!

bgilbert commented on rust/src/lib/query_apply/net_state.rs:
```diff
@@ -93,7 +94,9 @@ impl NetworkState {
         }
         if !self.kernel_only {
-            self.apply_with_nm_backend()
+            with_retry(RETRY_NM_INTERVAL_MILLISECONDS, RETRY_NM_COUNT, || {
```
It looks as though we're jumping directly to a two-second backoff?  That seems likely to be more than necessary, and would contribute to a meaningful boot delay.  Would it make sense to add more retries and reduce the interval between them?

cathay4t replied:
Retrying too frequent might lead to more race problems when user are creating 1000 VLANs. The retry here is more focusing on handle *unknown* problems. For this known problem, I have added a 2 seconds delay in `nmstatectl service` to avoid the meaningless retry. We may remove this sleep once NM fixed their problem.

bgilbert reviewed (commented):
The general backoff/retry approach SGTM.  I don't know this codebase so haven't tried to review in detail.

cathay4t committed:
nm: Retry when NetworkManager starting up

When OS boot up with `nmstate.service` enabled, nmstate will fail as
NetworkManager daemon is not fully setup the DBUS interface yet.

The `nmstate.service` has `After=NetworkManager.service`. Considering
`NetworkManager.service` is dbus type systemd service, it is guaranteed
that when `nmstate.service` starts, the DBUS interface is acquired by
NetworkManager. The failure is
`org.freedesktop.DBus.Error.UnknownMethod` as NetworkManager has not
finished the DBUS interface setup yet.

The `org.freedesktop.DBus.Error.UnknownMethod` is too generic, we cannot
assume it as daemon not started with checking the error string.
Considering the DBUS error string is not an API, we will not parse that
error message. Instead, we change top level retry to cover on querying
failures in `apply()`. The retry has been expend to twice with 2 seconds
interval(was 1 second interval).

The `nmstatectl service` command will sleep 2 seconds to avoid the
meaningless retry.

Do not want to create complex integration test case for this corner case.
Manually tested.

Signed-off-by: Gris Ge <fge@redhat.com>

ffmancera reviewed (approved):
I agree with the proposed workaround but I hope this could be fixed in NetworkManager so we do not need to do manual waiting of 2 seconds when getting an error..

Approving hoping we can drop this soon :)
